### DISCUSSION TITLE
I2S Sound Level Meter

### DISCUSSION THREAD

**Discussion by belveder79** [arendst/Tasmota#19427] [Ideas]
I would like to implement something like [this](https://github.com/ikostoski/esp32-i2s-slm) and I wonder if it is possible to do so, probably even with Berry? I compiled the current master with I2S input support at least...

**Comment by s-hadinger** [comment]
Currently, Berry only supports I2S output (Wav, MP3, Vorbis...). But there is no support for input yet.

Let's keep this alive though

**Comment by belveder79** [comment]
Ok, I see. I was wondering if I could just use the I2SBridge to send the data to a RaspPi and do some magic there... from reading the source and what the ESP32 occasionally tells me, it should send data over udp, so I could do something like

```bash
ffmpeg -i udp://192.168.0.116:6970 -acodec copy out.pcm
```

Unfortunately I can't get that to work and either I'm just doing something wrong, or the documentation seems to be broken somehow. I compiled the current master with

```
#ifndef I2S_BRIDGE
#define USE_I2S_AUDIO                       // Add support for I2S audio output
#define USE_I2S_MIC                         // Add support for I2S microphone
#define I2S_BRIDGE                          // Add support for UDP PCM audio bridge
  //#define I2S_BRIDGE_PORT    6970         // Set bridge port (default = 6970)
#endif
```
and flashed it. I can also see something like

```
I2S_bridge: command server created on port: 6971 
```

but I can't do any configuration on the cmd line. The command is simply not recognized?

```
17:41:01.503 CMD: I2S_bridge 0
17:41:01.506 SRC: WebConsole from 192.168.0.116
17:41:01.508 CMD: Grp 0, Cmd 'I2S_BRIDGE', Idx 1, Len 1, Pld 0, Data '0'
17:41:01.516 RSL: RESULT = {"Command":"Unknown"}
17:41:12.155 CMD: I2SBridge 0
17:41:12.158 SRC: WebConsole from 192.168.0.116
17:41:12.160 CMD: Grp 0, Cmd 'I2SBRIDGE', Idx 1, Len 1, Pld 0, Data '0'
17:41:12.167 RSL: RESULT = {"Command":"Unknown"}
```

I also tried to do some python3 command sending data, something like this:

```python
import socket;
UDP_IP="192.168.0.167"
UDP_PORT=6971
MESSAGE="cmd:0"
print("UDP target IP:", UDP_IP)
print("UDP target port:", UDP_PORT)
print("message:", MESSAGE)

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # UDP
sock.sendto(bytes(MESSAGE, "ascii"), (UDP_IP, UDP_PORT))
```

but that seems to crash the ESP32. Sometimes I can do send 0 and send 1 like this:

```
17:46:30.992 I2S_bridge: remote cmd 0
17:46:38.368 I2S_bridge: master started sending to ip: 192.168.0.116
17:46:38.371 I2S_bridge: remote cmd 1
```

but I'm receiving nothing.... there is no speaker, but I assigned the dout to the unused gpio 15...

<img width="350" alt="esp" src="https://github.com/arendst/Tasmota/assets/112880/a5ba900f-e90d-4e63-b702-c6232260f49c">


any help appreciated...

**Comment by Jason2866** [comment]
The i2s driver is heavily refactored and optimized for upcoming Tasmota version based on Arduino 3.0 / IDF5.1 
I2S API has changes in the underlying IDF 5.1 Current Arduino 2.0.x is built with IDF 4.4.5
Furthermore Berry support is/will be added. A sound level meter is planed as example.
Will not be done before Tasmota based on Arduino 3.0 has reached alpha state.
The used Arduino 3.0.0 is under heavy development and still experimental. It is in pre alpha state.
But it is already working nice and we don't expect not to many breaking changes there.
So alpha state may be soon reached for Arduino 3.0.0

**Comment by belveder79** [comment]
this is cool... I was looking at this coincidentally again today and at [this](https://dronebotworkshop.com/esp32-i2s/) one :-)

I have an I2S micro, INMP441, pretty standard. The only way to get things going is to use the I2SSay command once, which seems to set up the I2S overall... then Cmds 6, 4 and 2 in this order send over Mic Data, which can be played back with 

```ffplay -ar 32000 -ac 2 -i udp://192.168.0.116:6970```

but it is not good audio data - it is noise only as if the entire mic control is not properly working...

I know this is a WIP, so it's good to know that the level meter will be an example. Till then I'll likely try to follow the guide in this link see if the electronics work as expected at least...

**Reply by s-hadinger** [reply]
I'm working right now on the Berry Mic input mapping. I'm still struggling with the M5Stack Atom Echo, which has NS4168 I2S output (I2S 16 bits) and SPM1423 Mic (I2S PDM input). Both input and output share the same I2S WS clock, so I need to check if Tx/Rx work in full duplex mode (doc says that it can't) or dual-simplex mode with shared WS (which is unclear if it's possible)

**Comment by Jason2866** [comment]
@s-hadinger Strange that your device is not working. My Core2 did work fine with the driver from a few days from Staars. Mic (PDM) and Audio output.

**Comment by tablatronix** [comment]
I am also interested in this, my unifi cameras have no way to detect sound events like nest does, I want to add this to my tasmota nodes to at least have events which can trigger recording events etc..

**Reply by s-hadinger** [reply]
It's in the works...

**Reply by tablatronix** [reply]
Let me know if there is any experimentals you need testing on , I have a bunch of mems modules and will be diving into this in the next couple weeks

**Reply by s-hadinger** [reply]
Almost there. 

I tested with a PDM mems microphone from M5Stack Atom Echo. Quality is awful but it works. 

I needed to add a high pass filter to block the DC component that the PDM causes. I also needed to add a low pass filter to remove noise above 3Khz (it's configurable). 

Next the RMS works and provides a value every ~30ms. I'm looking for a simple UI to add a bar like vu meter.

**Reply by thomas-lentz** [reply]
@s-hadinger 
Did you get this to work? Is the RMS transmitted as TELE sensor-data? I was checking in the Tasmota documentation but couldn't find any hint yet.
I appreciate your help.
Greetings
 Tom
